mm2src/coins/utxo/utxo_common.rs

```diff
@@ -1749,44 +1749,101 @@ pub fn check_all_utxo_inputs_signed_by_pub(tx: &UtxoTx, expected_pub: &[u8]) ->
     Ok(true)
 }
 
+#[allow(clippy::too_many_arguments)]
 pub fn watcher_validate_taker_fee<T: UtxoCommonOps>(
     coin: T,
-    taker_fee_hash: Vec<u8>,
-    verified_pub: Vec<u8>,
+    taker_fee_hash: &[u8],
+    taker_payment_hex: &[u8],
+    output_index: usize,
+    sender_pubkey: &[u8],
+    amount: &BigDecimal,
+    min_block_number: u64,
+    fee_addr: &[u8],
+    lock_duration: u64,
 ) -> ValidatePaymentFut<()> {
+    let expected_amount = amount.clone();
+    let sender_pubkey = sender_pubkey.to_vec();
+    let taker_fee_hash = taker_fee_hash.to_vec();
+    let taker_payment_hex = taker_payment_hex.to_vec();
+    let address = try_f!(address_from_raw_pubkey(
+        fee_addr,
+        coin.as_ref().conf.pub_addr_prefix,
+        coin.as_ref().conf.pub_t_addr_prefix,
+        coin.as_ref().conf.checksum_type,
+        coin.as_ref().conf.bech32_hrp.clone(),
+        coin.addr_format().clone(),
+    )
+    .map_to_mm(ValidatePaymentError::InternalError));
```

Comment:
Please consider applying the following diff:
```diff
diff --git a/mm2src/coins/utxo/utxo_common.rs b/mm2src/coins/utxo/utxo_common.rs
index c8e715669..ce2d47f26 100644
--- a/mm2src/coins/utxo/utxo_common.rs
+++ b/mm2src/coins/utxo/utxo_common.rs
@@ -1765,18 +1765,9 @@ pub fn watcher_validate_taker_fee<T: UtxoCommonOps>(
     let sender_pubkey = sender_pubkey.to_vec();
     let taker_fee_hash = taker_fee_hash.to_vec();
     let taker_payment_hex = taker_payment_hex.to_vec();
-    let address = try_f!(address_from_raw_pubkey(
-        fee_addr,
-        coin.as_ref().conf.pub_addr_prefix,
-        coin.as_ref().conf.pub_t_addr_prefix,
-        coin.as_ref().conf.checksum_type,
-        coin.as_ref().conf.bech32_hrp.clone(),
-        coin.addr_format().clone(),
-    )
-    .map_to_mm(ValidatePaymentError::InternalError));
+    let fee_addr = fee_addr.to_vec();

     let fut = async move {
-        let expected_amount = sat_from_big_decimal(&expected_amount, coin.as_ref().decimals)?;
         let tx_from_rpc = match coin
             .as_ref()
             .rpc_client
@@ -1818,8 +1809,19 @@ pub fn watcher_validate_taker_fee<T: UtxoCommonOps>(
             )));
         }

+        let address = address_from_raw_pubkey(
+            &fee_addr,
+            coin.as_ref().conf.pub_addr_prefix,
+            coin.as_ref().conf.pub_t_addr_prefix,
+            coin.as_ref().conf.checksum_type,
+            coin.as_ref().conf.bech32_hrp.clone(),
+            coin.addr_format().clone(),
+        )
+        .map_to_mm(ValidatePaymentError::InternalError)?;
+
         match taker_fee_tx.outputs.get(output_index) {
             Some(out) => {
+                let expected_amount = sat_from_big_decimal(&expected_amount, coin.as_ref().decimals)?;
                 let expected_script_pubkey = Builder::build_p2pkh(&address.hash).to_bytes();
                 if out.script_pubkey != expected_script_pubkey {
                     return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
```

It's better to compute values when we can can use them. In current implementation, there are multiple error handling steps that might raise error. In that case those computations will be pointless.